README.md
# Area58check

## What is Area58?

Area58 is a library to encode and decode base58check. Base58check is
used in Bitcoin whenever there is a need for a user to read or
transcribe a number, such a bitcoin adresses, encrypted key, private
key, or script hash.

## What is NOT Area58?

This can't be used to generate bitcoin address from a private key. It's
only a base58check encoder/decoder. But it's easy to generate a bitcoin
address from the private key.

Example: If you want to generate a (uncompressed) bitcoin address using
this library you can...

    iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
    iex> {uncompressed_pubkey, _priv_key} = :crypto.generate_key(:ecdh, :crypto.ec_curve(:secp256k1), privkey)
    iex> uncompressed_pubkey = :crypto.hash(:ripemd160, :crypto.hash(:sha256, uncompressed_pubkey))
    iex> Area58check.encode(uncompressed_pubkey, version: <<0>>)
    %{encoded: "1CLrrRUwXswyF2EVAtuXyqdk4qb8DSUHCX",
      version: :p2pkh,
      version_bin: <<0>>}

or you can use another library that does all that for you.

## What is base58check?

#### Encoding
Base58 charset is based on easy to trancribe characters, upper and
lower case letters and numbers excluding `0`, `O`, `l`, `I`.

#### Bitcoin version prefixes
Usually there is a version (binary number) to signal what kind of string
is being encoded.
[list of bitcoin prefixes](https://en.bitcoin.it/wiki/List_of_address_prefixes)

Example: for standard bitcoin address, is used the hexadecimal
number 0x00, for compressed WIF (allows to import a private key in
bitcoin) is used 0x80, etc.

#### Checksum

When a string is encoded, 4 bytes (or 32bits), are added to the end by
`Area58check.encode` function as a checksum to make sure that no
characters were wrongly transcribed. This checksum is not meant to allow
to fix a wrong transcribed string, but it allows to verify if string is
correct or not.

## Usage

#### Encoding

Encode a private key into WIF (Wallet Import Format):

    iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
    iex> Area58check.encode(privkey, version: :wif )
    %{encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
      version: :wif,
      version_bin: <<128>>}
    iex> Area58check.encode(privkey, version: [128] )
    %{encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
      version: :wif,
      version_bin: <<128>>}
    iex> Area58check.encode(privkey, version: <<128>>)
    %{encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
      version: :wif,
      version_bin: <<128>>}
    iex> Area58check.encode(privkey, version: 0x80)
    %{encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
      version: :wif,
      version_bin: <<128>>}

Encode a public key into an address:

    iex> {uncompressed_pubkey, _priv_key} = :crypto.generate_key(:ecdh, :crypto.ec_curve(:secp256k1), privkey)
    iex> derived_uncomp_pubkey = :crypto.hash(:ripemd160, :crypto.hash(:sha256, uncompressed_pubkey))
    iex> Area58check.encode(derived_uncomp_pubkey, version: 0x00)
    %{encoded: "1CLrrRUwXswyF2EVAtuXyqdk4qb8DSUHCX",
      version: :p2pkh,
      version_bin: <<0>>}

Error when version is unknown:

    iex> Area58check.encode(privkey, version: :jhkdsfajhkfdasjhkfd)
    ** (ArgumentError) :jhkdsfajhkfdasjhkfd is not a recognized version.
    You can either pass a charlist (ex: [4, 136, 178, 30]), number (ex:
    70617039), hexadecimal (ex: 0x043587CF), binary version (ex: <<4,
    136, 178, 30>>), or a recognized atom like any of: :bip32_privkey,
    :bip32_pubkey, :p2pkh, :p2sh, :tesnet_bip32_privkey,
    :tesnet_bip32_pubkey, :tesnet_p2pkh, :tesnet_p2sh, :tesnet_wif, :wif

#### Decoding

Decoding strings encoded with base58check:

    iex> Area58check.decode("1CLrrRUwXswyF2EVAtuXyqdk4qb8DSUHCX")
    {:ok,
      %{decoded: <<124, 106, 230, 190, 9, 150, 81, 133, 169, 75, 13, 161, 139, 201,
          42, 157, 252, 238, 97, 23>>,
        version: :p2pkh,
        version_bin: <<0>>}}
    iex> Area58check.decode("5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb")
    {:ok,
      %{decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205,
          239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205,
          239>>,
        version: :wif,
        version_bin: <<0>>}}

If checksum is not valid returns error:

    iex> Area58check.decode("1CheckSumError")
    {:error, :checksum_incorrect}

If contains a non valid character like `0`, `O`, `l`, `I`, returns
error:

    iex> Area58check.decode("ContainsInvalidCharacter0")
    {:error, :incorrect_base58}

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `area58check` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:area58check, "~> 0.1"},
  ]
end
```

The docs can be found at
[https://hexdocs.pm/area58check](https://hexdocs.pm/area58check).



CHANGELOG.md
## v0.1.0 (2018-04-03)

First release


test/area58check_test.exs
defmodule Area58checkTest do
  use ExUnit.Case
  doctest Area58check, import: true

  describe "Encode and decode public key" do
    @pubkeyhash <<124, 106, 230, 190, 9, 150, 81, 133, 169, 75, 13, 161, 139, 201, 42, 157, 252, 238, 97, 23>>
    test "" do
      %{encoded: encoded} = Area58check.encode(@pubkeyhash, :p2pkh)
      {:ok, %{decoded: decoded}} = Area58check.decode(encoded)
      assert decoded === @pubkeyhash
    end
  end

  describe "Encode and decode a random binary" do
    # Generate a random binary of length ranging from 0 to 50 bytes.
    @random_bin :crypto.strong_rand_bytes(Enum.random(0..50))
    @version 0
    test "will be the same string" do
      %{encoded: encoded} = Area58check.encode(@random_bin, @version)
      {:ok, %{decoded: decoded}} = Area58check.decode(encoded)
      assert decoded === @random_bin
    end
  end

end


test/area58check/crypto_test.exs
defmodule Area58checkTest.CryptoTest do
  use ExUnit.Case
  doctest Area58check.Crypto, import: true
end


test/area58check/decoder_test.exs
defmodule Area58checkTest.DecoderTest do
  use ExUnit.Case
  doctest Area58check.Decoder, import: true
end


mix.exs
defmodule Area58check.Mixfile do
  use Mix.Project

  @github_url "https://github.com/ihinojal/area58check"

  def project do
    [
      app: :area58check,
      version: "0.1.0", # Major: Incompatible API changes
                        # Minor: Backward-compatible added functionality
                        # Patch: Backward-compatible bug fixes
      elixir: "~> 1.5",
      description: description(),
      package: package(),
      start_permanent: Mix.env == :prod,
      deps: deps(),
      source_url: @github_url,
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  #── PRIVATE ──────────────────────────────────────────────────────────

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:dialyxir, "~> 0.5", only: [:dev], runtime: false},
      {:ex_doc, "~> 0.16", only: :dev, runtime: false},
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"},
    ]
  end

  defp description do
    "Base58check library to encode binary data or decode encoded "<>
    "strings. It's used in Bitcoin whenever there is a need for a "<>
    "user to read or transcribe a number, such a bitcoin adresses, "<>
    "encrypted key, private key, or script hash."
  end

  defp package do
    [maintainers: ["Ivan H."],
     licenses: ["MIT"],
     links: %{"Github" => @github_url}]
  end

  defp docs do
    [
      main: "Area58check",
      # source_ref: "v#{@version}",
      # logo: "path/to/logo.png",
      extras: ["README.md"]
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :area58check, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:area58check, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/area58check/crypto.ex
defmodule Area58check.Crypto do
  @moduledoc false
  #Several crypto related functions in erlang and wrapped in elixir
  #functions.

  @doc """
  sha256 hash function.

  It always returns a binary of length 64 Bytes (or 256 bits).

  ## Examples:

  It works in any binary argument:

    iex> sha256("Any binary string")
    <<234, 79, 243, 250, 95, 195, 243, 27, 227, 50, 104, 90, 40, 117, 226, 203, 9, 226, 41, 30, 162, 39, 57, 180, 192, 110, 151, 230, 15, 185, 157, 229>>

  But it's mainly used when working with public keys:

    iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
    iex> {uncompressed_pubkey, _priv_key} = :crypto.generate_key(:ecdh, :crypto.ec_curve(:secp256k1), privkey)
    iex> _pubkey = ripemd160(sha256(uncompressed_pubkey))
    <<124, 106, 230, 190, 9, 150, 81, 133, 169, 75, 13, 161, 139, 201, 42, 157, 252, 238, 97, 23>>
  """
  @spec sha256(String.t) :: <<_::256>>
  def sha256(payload) do
    :crypto.hash(:sha256, payload)
  end

  @doc """
  ripemd160 hash function.

  It always returns a binary of length 160 bits (or 20 Bytes).

  ## Examples:

  It works in any binary argument:

    iex> ripemd160("Any binary string")
    <<151, 23, 122, 58, 30, 171, 33, 75, 84, 65, 91, 222, 73, 172, 26, 56, 95, 190, 26, 236>>

  But it's mainly used when working with public keys:

    iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
    iex> {uncompressed_pubkey, _priv_key} = :crypto.generate_key(:ecdh, :crypto.ec_curve(:secp256k1), privkey)
    iex> ripemd160(sha256(uncompressed_pubkey))
    <<124, 106, 230, 190, 9, 150, 81, 133, 169, 75, 13, 161, 139, 201, 42, 157, 252, 238, 97, 23>>
  """
  @spec ripemd160(String.t) :: <<_::160>>
  def ripemd160(payload) do
    :crypto.hash(:ripemd160, payload)
  end

  @doc """
  Base58check checksum used in base58check. It will be attached so we
  can later check if any of the characters of the payload has been
  altered.

  It always return 4 Bytes.

  The checksum in base58check is defined by:
    1. Compute 64 Bytes hash of the payload.
    2. Compute 64 Bytes hash of step 1.
    3. Take only the first 4 bytes of 2.

  ## Examples:

  It works in any binary argument:

    iex> checksum("Any binary string")
    <<129, 119, 29, 167>>

  Also works with charlist:

    iex> checksum(<<1,2,255>>) === checksum([1,2,255])
    true

  But it's mainly used with public or private keys:

    iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
    iex> version_prefix = <<0>>
    iex> checksum(version_prefix <> privkey)
    <<2, 252, 125, 189>>
  """
  @spec checksum(String.t | [byte]) :: <<_::64>>
  def checksum(payload) do
    payload
    |>sha256() # Return a 64 Bytes hash
    |>sha256() # Return a second 64 Bytes re-hash
    |>Kernel.binary_part(0, 4) # Take only first 4 bytes
  end

end


lib/area58check/decoder.ex
defmodule Area58check.Decoder do
  @moduledoc false
  # This module allows to decode a base58check binary to a map which
  # tells if it's correct.

  alias Area58check.{Crypto, Base58symbolChart, Prefixes}

  @doc """
  Decode a base58check binary to a map which tells if it's correct, and
  the type of string comparing the prefix to the already known prefixes.

  ## Examples:

      iex> {:ok, %{decoded: decoded}} = decode_string("5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb")
      {:ok, %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version_bin: <<128>>,
        version: :wif
      }}
      iex> _privkey = decoded |> Base.encode16()
      "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF"

  It checks the checksum to verify if correct.

      iex> decode_string("")
      {:error, :checksum_incorrect}
      iex> decode_string("5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWc")
      {:error, :checksum_incorrect}

  Check if alphabet is correct. `O` is not a valid base58 character.

      iex> decode_string("Ol")
      {:error, :incorrect_base58}
  """
  @spec decode_string(String.t) :: {:ok, Area58check.t} | {:error, atom}
  def decode_string(string) do
    string # "1zhML"
    |>Base58symbolChart.check_valid_base58_chars() #=> "1zhML" or {:error, :incorrect_base58}
    |>decode_raw_string() #=> [0, 57, 40, 20, 19] or {:error, reason}
    |>verify_checksum() #=> %{:ok, %{decoded: <<..>>} or {:error, reason}
    |>extract_version() #=> %{:ok, %{decoded: <<..>>, ver...} or {:error, reason}
    |>merge_encoded(string) #=> %{:ok, %Area58check{encoded: ..., decoded: <<..>>, ver...} or {:error, reason}
  end

  #── PRIVATE ──────────────────────────────────────────────────────────
  # Merge the current result with the original encoded string
  defp merge_encoded({:ok, result}, encoded_str) do
    {:ok, Map.merge(%Area58check{encoded: encoded_str}, result)}
  end
  defp merge_encoded({:error, reason}, _string), do: {:error, reason}

  # if string starts with "1", it means that the real decoded value is
  # 0. We need to convert it separatelly or otherwise it will be lost by
  # the `Integer.undigits/2` function
  # Ex: decode_raw_string("1zhML")
  defp decode_raw_string("1"<>rest) do
    [Base58symbolChart.ascii_to_base58_char("1") | decode_raw_string(rest)]
  end
  # Ex: decode_raw_string("zhML")
  defp decode_raw_string(vers_payload_checksum) when is_binary(vers_payload_checksum) do
    vers_payload_checksum # => "zhML"
    # Convert back to integers in base58 from base58 characters
    |>Base58symbolChart.ascii_to_base58() #=> [57, 40, 20, 19]
    # Convert to a big endian number
    |>Integer.undigits(58) # 11_257_123
    # Convert that number to a list of base 256 (binary)
    |>Integer.digits(256) # [171, 197, 35]
  end
  # If there was an error previously just bubble up to the next piped
  # function
  defp decode_raw_string({:error, reason}), do: {:error,reason}

  # If there was an error previously just pass it to the next piped
  # function
  defp verify_checksum({:error, map}) do
    {:error, map}
  end
  # Ex: verify_checksum([0, 57, 40, 20, 19])
  defp verify_checksum(vers_payload_checksum) do
    # Split the version and payload from the checksum
    {vers_payload, checksum} =
      vers_payload_checksum
      |>Enum.split(-4) #=> {[4,16,47,...], [43,52,29,34]}
    # Calculate checksum and compare with provided to verify is correct
    if Crypto.checksum(:binary.list_to_bin(vers_payload)) === :binary.list_to_bin(checksum) do
      {:ok, vers_payload} #=> {:ok, [4,16,47,...]}
    else
      {:error, :checksum_incorrect}
    end
  end

  # From the binary which contains the version and the payload, try to
  # extract the version
  #Ex: extract_version({:ok, [0, 57, 40, 20, 19]})
  defp extract_version({:ok, vers_payload}) do
    vers_payload #=> [0, 57, 40, 20, 19]
    |>:binary.list_to_bin() #=> <<0, 57, 40, 20, 19>>
    |>Prefixes.get_version() #=> {:ok, %{decoded: ..., version: :p2pkh, ...}}
  end
  defp extract_version({:error, reason}), do: {:error, reason}

end


lib/area58check/encoder.ex
defmodule Area58check.Encoder do
  @moduledoc false
  # This module allows to encode a binary to a base58check.

  alias Area58check.{Crypto, Base58symbolChart}

  @doc """
  Encode a binary into base58check with a version of the type of string
  being encoded adding a checksum.

  Arguments is a binary string of any size (normally a 64Bytes is used),
  plus the version also in binary.

  ## Examples

    iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
    iex> {uncompressed_pubkey, _priv_key} = :crypto.generate_key(:ecdh, :crypto.ec_curve(:secp256k1), privkey)
    iex> derived_uncomp_pubkey = :crypto.hash(:ripemd160, :crypto.hash(:sha256, uncompressed_pubkey))
    iex> encode_string(derived_uncomp_pubkey, <<0>>)
    "1CLrrRUwXswyF2EVAtuXyqdk4qb8DSUHCX"

  Encode null string:

    iex> encode_string("", <<0>>)
    "1Wh4bh"
  """
  @spec encode_string(binary, binary) :: String.t | no_return()
  def encode_string(payload, version_prefix) when is_binary(version_prefix) do
    # Concatenate the version with the string to be encoded
    versioned_payload = version_prefix <> payload # <<0>> <> <<1, 2, 3>>
    # Concatenate the (version+payload + checksum) together
    (versioned_payload <> Crypto.checksum(versioned_payload))
    |>encode_versioned_string()
  end
  def encode_string(_payload, _version_prefix) do
    raise ArgumentError, message: "First argument should be a binary"
  end

  #── PRIVATE ──────────────────────────────────────────────────────────

  # Encode string assuming that the version, payload and checksum are
  # concatenated together.

  # If string starts with zero, those will be lost if we don't convert
  # separately.
  # Ex: encode_versioned_string(<<0,1,2,3>>)
  defp encode_versioned_string(<<0, checksumed_payload::binary>> ) do
    Base58symbolChart.base58_to_ascii_char(0) <>
      encode_versioned_string(checksumed_payload)
  end
  # This asumes that `vers_payload_check` contains concatenated version_prefix <>
  # payload <> checksum. Payload should not contain any leading zeros or
  # they will be lost.
  # Ex: encode_versioned_string(<<1,2,3>>)
  defp encode_versioned_string(vers_checksumed_payload)
    when is_binary(vers_checksumed_payload) do
    vers_checksumed_payload #=> <<124, 106, 230, 190, 9, 150, 81, 133, 169, 75, 13, 161, 139, 201, 42, 157, 252, 238, 97, 23, 202, 56, 155, 228>>
    # Convert to integer as a single big-endian bignumber. Should not
    # contain any leading zeros, as they will discarded.
    |>:binary.decode_unsigned(:big) #=> 3050710266992610275399901494860636957854608474636618865636
    # Convert that number to a list of base 58 number. Change big
    # decimal number to base 58.
    |>Integer.digits(58) #=> [11, 19, 49, 49, 24, 27, 54, 30, 50, 54, 56, 14, 1, 13, 28, 9, 51, 52, 30, 56, 48, 36, 43, 3, 48, 34, 7, 12, 25, 27, 16, 11, 30]
    # Convert each byte of list to a byte of the base58 according the
    # base58 character table. Ex: '0' becomes '1', '25' becomes 'S'
    |>Base58symbolChart.base58_to_ascii() #=> "CLrrRUwXswyF2EVAtuXyqdk4qb8DSUHCX"
  end

end


lib/area58check.ex
defmodule Area58check do
  @moduledoc """
  This module includes the needed functions to encode or decode
  base58check strings.

  Base58check is used in Bitcoin whenever there is a need for a user to
  read or transcribe a number, such a bitcoin adresses, encrypted key,
  private key, or script hash.

  ## Prefixed versions
  The prefixed version parameter indicates what kind of binary is being
  encoded. E.g. If you are encoding a private key use `:wif`, if you are
  encoding a bitcoin address use `:p2pkh`. The version prefix will be
  embeded in the encoded version.
  You can also use other types described in `t:prefix_version_type/0`

  ## Checksum
  The encoded versions will include a few bytes to ensure that its
  correct and no transcribing errors were present.

  ## Usage

  To know the **decoded binary** of any encoded string just use the
  `encode/2` function indicating the prefix version you want:

      iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
      iex> %{decoded: _decoded} = encode(privkey, <<128>>)
      %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>
      }

  To know the **encoded string** of any binary use `decode/1` function:

      iex> {:ok, %{encoded: _encoded}} = decode("5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb")
      {:ok, %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>
      }}
  """
  alias __MODULE__.{Prefixes, Encoder, Decoder}

  @typedoc """
  Indicates the type of bitcoin string being encoded/decoded. E.g.: mainet
  WalletImportFormat (WIF), bitcoin address, testnet wif, etc.

  The prefixed version can be provided in several formats:

    - An atom. E.g.: `:testnet_p2sh`
    - The hexadecimal or numeric representation. E.g.: `0x043587CF`
    - Character list: E.g.: `[4, 136, 178, 30]`
    - Binary string: E.g.: `<<4, 136, 178, 30>>`
  """
  @type prefix_version_type :: atom | binary | pos_integer | nonempty_list(0..255)
  @typedoc """
  A result returned by both `encode/2` and `decode/1` with the prefixed
  version, the encoded base58check string, and the decoded binary.
  """
  @type t :: %Area58check{
    decoded: binary,
    encoded: String.t,
    version: atom,
    version_bin: binary}

  defstruct decoded: <<>>, encoded: "", version: nil, version_bin: <<>>

  @doc """
  Encodes a string using base58check with the provided version prefix .

  The argument `version_prefix` can be a binary, integer or a list of
  bytes (see examples next paragraph).
  All available prefixes are coded in the module `Area58check.Prefixes`.

  ## Examples

      iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
      iex> encode(privkey, :wif)
      %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>}

  The version prefix can be indicated in several alternative ways:

      iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
      iex> encode(privkey, <<128>>)
      %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>
      }
      iex> encode(privkey, 0x80)
      %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>
      }
      iex> encode(privkey, [128])
      %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>
      }

  Example to generate an uncompresed bitcoin address from public key

      iex> privkey = "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF" |> Base.decode16!()
      iex> {uncompressed_pubkey, _priv_key} = :crypto.generate_key(:ecdh, :crypto.ec_curve(:secp256k1), privkey)
      iex> uncompressed_pubkey = :crypto.hash(:ripemd160, :crypto.hash(:sha256, uncompressed_pubkey))
      iex> encode(uncompressed_pubkey, <<0>>)
      %Area58check{
        encoded: "1CLrrRUwXswyF2EVAtuXyqdk4qb8DSUHCX",
        decoded: <<124, 106, 230, 190, 9, 150, 81, 133, 169, 75, 13, 161, 139, 201, 42, 157, 252, 238, 97, 23>>,
        version: :p2pkh,
        version_bin: <<0>>}

  If version is an atom, and is not recognized returns error:

      iex> encode("Any string", :sahdkjfhkjasdfhksldjf)
      ** (ArgumentError) Version prefix :sahdkjfhkjasdfhksldjf is not a recognized version. You can either pass a charlist (ex: [4, 136, 178, 30]), number (ex: 70617039), hexadecimal (ex: 0x043587CF), binary version (ex: <<4, 136, 178, 30>>), or a recognized atom like any of: :bip32_privkey, :bip32_pubkey, :p2pkh, :p2sh, :tesnet_bip32_privkey, :tesnet_bip32_pubkey, :tesnet_p2pkh, :tesnet_p2sh, :tesnet_wif, :wif
  """
  @spec encode(String.t, prefix_version_type) :: t
  def encode(payload, version_prefix ) do
    # E.g.: get_binary_version(:pubkey_hash) # => {<<0>>}
    {version, version_bin} = Prefixes.get_binary_version( version_prefix )
    %Area58check{
      encoded: Encoder.encode_string(payload, version_bin),
      decoded: payload,
      version: version,
      version_bin: version_bin}
  end

  @doc """
  Decodes a previously encoded string using base58check returning the
  used version prefix, the original binary.

  An error message will be returned in case there was a checksum error
  or a character is not valid.

  ## Examples:

  Decode WIP key and get the hexadecimal private key

      iex> {:ok, %{decoded: decoded} } = decode("5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb")
      {:ok, %Area58check{
        encoded: "5HpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb",
        decoded: <<1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239, 1, 35, 69, 103, 137, 171, 205, 239>>,
        version: :wif,
        version_bin: <<128>>}}
      iex> decoded |> Base.encode16
      "0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF"

  If any character of the key is wrong (`J`), checksum will be invalid:

      iex> decode("5JpneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb")
      {:error, :checksum_incorrect}

  The null string means the checksum is nor present, nor valid:

      iex> decode("")
      {:error, :checksum_incorrect}

  If any character is not a base58 character (E.g.: `0`)

      iex> decode("50pneLQNKrcznVCQpzodYwAmZ4AoHeyjuRf9iAHAa498rP5kuWb")
      {:error, :incorrect_base58}
  """
  @spec decode(String.t) :: {:ok, Area58check.t} | {:error, atom()}
  def decode(payload) do
    Decoder.decode_string(payload)
  end

end
